akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -426,104 +424,273 @@ object StreamLayout {
-            case _ ⇒
-              tryOnSubscribe(sub, wrapped)
-              wrapped.closeLatch() // allow onNext only now
-              terminationStatus.set(Allowed)
+  override final def onSubscribe(s: Subscription): Unit = {
+    @tailrec def rec(obj: AnyRef): Unit =
+      get() match {
+        case null => if (!compareAndSet(null, obj)) rec(obj)
+        case subscriber: Subscriber[_] =>
+          obj match {
+            case subscription: Subscription =>
+              if (compareAndSet(subscriber, Both.create(subscriber))) establishSubscription(subscriber, subscription)
+              else rec(obj)
+            case pub: Publisher[_] =>
+              getAndSet(Inert) match {
+                case Inert => // nothing to be done
+                case _     => pub.subscribe(subscriber.asInstanceOf[Subscriber[Any]])
+              }
           }
-        case sub: Subscription ⇒
-          s.cancel() // reject further Subscriptions
+        case _ =>
+          // spec violation
+          tryCancel(s)
       }
+
+    if (s == null) {
+      val ex = subscriptionMustNotBeNullException
+      try rec(ErrorPublisher(ex, "failed-VirtualProcessor"))
+      finally throw ex // must throw NPE, rule 2:13
+    } else rec(s)
   }
 
-  override def onError(t: Throwable): Unit = {
-    requireNonNullException(t)
-    if (terminationStatus.compareAndSet(null, Failed(t))) () // let it be picked up by subscribe()
-    else tryOnError(subscriptionStatus.get.asInstanceOf[Subscriber[T]], t)
+  private def establishSubscription(subscriber: Subscriber[_], subscription: Subscription): Unit = {
+    val wrapped = new WrappedSubscription(subscription)
+    try subscriber.onSubscribe(wrapped)
+    catch {
+      case NonFatal(ex) =>
+        set(Inert)
+        tryCancel(subscription)
+        tryOnError(subscriber, ex)
+    }
   }
 
-  override def onComplete(): Unit =
-    if (terminationStatus.compareAndSet(null, Completed)) () // let it be picked up by subscribe()
-    else tryOnComplete(subscriptionStatus.get.asInstanceOf[Subscriber[T]])
+  override def onError(t: Throwable): Unit = {
+    @tailrec def rec(ex: Throwable): Unit =
+      get() match {
+        case null =>
+          if (!compareAndSet(null, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec(ex)
+          else if (t == null) throw ex
+        case s: Subscription =>
+          if (!compareAndSet(s, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec(ex)
+          else if (t == null) throw ex
+        case Both(s) =>
+          set(Inert)
+          try tryOnError(s, ex)
+          finally if (t == null) throw ex // must throw NPE, rule 2:13
+        case s: Subscriber[_] => // spec violation
+          getAndSet(Inert) match {
+            case Inert => // nothing to be done
+            case _     => ErrorPublisher(ex, "failed-VirtualProcessor").subscribe(s)
+          }
+        case _ => // spec violation or cancellation race, but nothing we can do
+      }
```

Comment:
These also look correct. I would prefer an extra comment about the treatment of `null`s here.